### DISCUSSION TITLE
Improve usability of theme typings when using `theme only` approach to style everything

### DISCUSSION THREAD

**Discussion by TheThirdRace** [chakra-ui/chakra-ui#3568] [General]
# Goal of this discussion

Improve the usability of theme typings when using only the theme to style everything.

I hope this discussion can shed some light on how `theme only` styling is used in the wild and what could be done to ease the pain points encountered.

⠀

## Intro

The most popular way to use Chakra-ui is to use `styling props` directly on the components. It's very easy to understand and apply this way. As far as I know, typings are working just fine for this method.

A less popular way to use Chakra-ui is to use a `theme only` approach to style everything. The components only use `useStyleConfig`, `useMultiStyleConfig` and the `sx` property.

It is a more complexed method than using `styling props`, but using a `theme only` approach to style everything has multiple advantages over the popular method:

- Components are reusable between projects without any modification
- Separation of concerns: logic is in JS while styling is in theme
- If you ever want to change the style completely, you only have to check the theme folder
- It's entirely possible to switch themes on the fly, not just colors, making total conversions a possibility at the flip of a switch (see http://www.csszengarden.com/)

⠀

## Custom typings I currently use

Whenever I encountered a situation I couldn't simply import a type I needed from Chakra, I literally opened the source and copied the typings in a local file.

Even after copying the typings, I still had to make some adjustments to get around some pain points.

It's important to note those typings were done before the new typings from version `1.3.2`.

```typescript
import { ColorMode, SystemStyleObject, Theme, ThemingProps } from '@chakra-ui/react'

export type StyleConfig = {
	parts?: string[]
	baseStyle: StyleInterpolation
	sizes?: { [size: string]: StyleInterpolation }
	variants?: { [variant: string]: StyleInterpolation }
	defaultProps?: {
		size?: string
		variant?: string
	}
}

export type StyleConfigProps = {
	theme: Theme
	colorMode: ColorMode
	colorScheme: string
}

export type StyleConfigFunction =
	| ((props: StyleConfigProps) => SystemStyleObject)
	| ((props: StyleConfigProps) => { [part: string]: SystemStyleObject })

type StyleInterpolation = SystemStyleObject | { [part: string]: SystemStyleObject } | StyleConfigFunction
```

⠀
⠀
⠀

## `#1` Passing `props` to a function (part 1)

There are times when you need to consolidate some logic for styling.

`textStyles` and `layerStyles` are fulfilling this role for a `styling props` approach, but until very recently, you couldn't use them in a `theme only` approach.

To get around this, you could create a function and pass the props you'd receive instead.

```typescript
// typings for component Text
import { SystemStyleObject } from '@chakra-ui/react'
import { StyleConfig, StyleConfigProps } from '../../type/Chakra'
import { text } from '../variables/brand'

export const Text: StyleConfig = {
	baseStyle: {},
	variants: {
		highlight: (props: StyleConfigProps): SystemStyleObject => ({
			color: text.highlight(props), // <--- See how props is passed to text.highlight()
			textDecoration: 'line-through'
		}),
	},
	defaultProps: {}
}
```

```typescript
// text.highlight() function
export const text = {
	highlight: (props: StyleConfigProps): string =>
		`linear-gradient(130deg,${getColor(props.theme, 'primary.500')},${getColor(
			props.theme,
			'secondary.500'
		)} 40%,${getColor(props.theme, 'secondary.1000')} 75%)`,
}
```

Now, this works 100% with my custom typings.

But I can't do this kind of things at all with Chakra's original typings, neither with typings from `1.3.2` onward.

Why? Simply because the typings are not exported for `props`. This makes Typescript very uncooperative when passing stuff around...

Here's how typings for `props` are defined in `1.3.2`:

```typescript
export declare type ThemingPropsThunk<T> = T | ((props: Dict & Omit<ThemingProps, "styleConfig"> & {
		 colorMode: ColorMode;
	}) => T);

export interface ComponentSingleStyleConfig {
    baseStyle?: ThemingPropsThunk<SystemStyleObject>;
    sizes?: Record<string, ThemingPropsThunk<SystemStyleObject>>;
    variants?: Record<string, ThemingPropsThunk<SystemStyleObject>>;
    defaultProps?: ComponentDefaultProps;
}
```

The part I need for a function receiving `props` is this:

```typescript
Dict &
	Omit<ThemingProps, 'styleConfig'> & {
		colorMode: ColorMode
	}
```

I can't import the type for `props` directly from Chakra because it's inlined and not a separate entity.

The only workaround I have at the moment is to create a new type locally and copy/paste the definition from Chakra. The obvious downside to this method is now I have to manually change the definition whenever Chakra changes it's own.

To solve this pain point, I would need a separate type exported by Chakra with this inside

```typescript
export type StyleConfigProps = Dict &
			Omit<ThemingProps, 'styleConfig'> & {
				colorMode: ColorMode
			}

export declare type ThemingPropsThunk<T> = T | ((props: StyleConfigProps) => T);

export interface ComponentSingleStyleConfig {
    baseStyle?: ThemingPropsThunk<SystemStyleObject>;
    sizes?: Record<string, ThemingPropsThunk<SystemStyleObject>>;
    variants?: Record<string, ThemingPropsThunk<SystemStyleObject>>;
    defaultProps?: ComponentDefaultProps;
}
```

⠀
⠀
⠀

## `#2` Passing `props` to a function (part 2)

Even with the exported type from pain point #1, I'm still missing something.

The new typing from `1.3.2` doesn't include the property `theme` in `props`, which means auto-suggest doesn't list `theme` as an acceptable value to use anymore.

I can definitely reference `props.theme` even if it's not auto-suggested, but `theme` now also has the type `any`.

This has 2 unfortunate consequences:

1. `props.theme.` doesn't give any auto-suggest for any property down the chain
2. I'm not type safe anymore

While my custom typings weren't perfect, they would at least give me the typings of the default Chakra theme. (more on this just below)

To solve this, I need `theme` to be suggested as a possible property and with the right typings too.

⠀
⠀
⠀

## `#3` `props.theme` can't have typings from extensions to the default theme

When using `styling props`, you can define your theme and overrides, then define its type:

```typescript
const theme = extendTheme(overrides)
export default theme
export type Theme = typeof(theme)
```

With this, it's very easy to use hooks and get the right typings.

But in `theme only` approach, I cannot use that type in the extensions because it creates a circular dependency.

Example:

```typescript
// typings for button
//    which doesn't really make sens, but get the point across
import { Theme } from '../index.ts' // <--- This is the `typeof(theme)`
export const Button: ComponentStyleConfig = {
	baseStyle: {},
	variants: {
		drawerClose: (props) => {
			const outline = (props.theme as Theme).components.Button.variants.outline(props)
			return {
				...outline,
				fontSize: 'inherit',
				fontWeight: 'normal'
			}
		},
		outline: (props) => ({
			borderColor: borderColor.base(props),
			color: 'inherit'
		}),
	},
	defaultProps: {}
}
```

In this case, the type `Theme` can only be created once all the theme is settled, but in `Button` override, it needs the `Theme` to settle its types correctly.

The circular dependency is preventing me from having complete theme typings.

The only way I can get around this is to change the code like this:

```typescript
// typings for button
//    which doesn't really make sens, but get the point across
export const Button: ComponentStyleConfig = {
	baseStyle: {},
	variants: {
		drawerClose: (props) => {
			const outline = (props.theme.components.Button as ButtonExtended).variants.outline(props)
			return {
				...outline,
				fontSize: 'inherit',
				fontWeight: 'normal'
			}
		},
		outline: (props) => ({
			borderColor: borderColor.base(props),
			color: 'inherit'
		}),
	},
	defaultProps: {}
}

export type ButtonExtended = typeof(Button)
```

I tried to use the new `Chakra CLI` to see if I could generate the typings correctly, but documentation is a bit spotty. It doesn't say what am I supposed to import once the theme typings have been generated.

Also the current `framer-motion` bug is casting a lot of doubts if the CLI is even working at all...

```
× An error occurred
motion.custom() is deprecated. Use motion() instead.
```

In any case, from what I could understand, the `Chakra CLI` is supposed to generate the whole theme typings in `node_modules`.

If I'm not mistaken, this means everytime I create a new variant or a new `part` for a component, I would have to run the CLI manually so I can get the typings right.

If I'm right, it's still less than ideal... But at least it would be an official workaround.

I'm still pondering on this, but I think my current solution might be better than `Chakra CLI` at this point. With my current implementation, I need more code, but I instantly know if there's a problem in the typings, while the `CLI` would let me believe everything's fine until I run it again to generate the new typings and then see the errors.

⠀
⠀
⠀

## `#4` Inferred typings for returned values

I always have a hard time accepting inferred typing for return values. Especially when dealing with complex objects that should respect a specific format.

```typescript
import { ComponentMultiStyleConfig } from '@chakra-ui/react'
import { background, borderColor, darkTheme, text } from '../variables/brand'
import { edgeMargins, footerMaxWidth, footerMinWidth } from '../variables/page'

export const PageFooter: ComponentMultiStyleConfig = {
	parts: ['layParent', 'layChild', 'preChild'],
	baseStyle: {
		layParent: {},
		layChild: {},
		preChild: {}
	},
	variants: {
		default: (props) => ({
			layParent: {},
			layChild: {
				display: 'flex',
				flexDirection: 'column',
			},
			preChild: {}
		})
	},
	defaultProps: {}
}
```

In this example, I could easily veer off the path and start to define `baseStyle` like this:

```typescript
baseStyle: {
		layParent: {},
		layChild: {
			brand1: {
				color: 'red.100'
			},
			brand2: {
				subbrand1: {
					color: 'blue.100'
				},
				subbrand2: {
					color: 'green.100'
				}
			}
		},
		preChild: {}
	},
```

With the new typings I can structure my theme in all kinds of wrong ways.

And Typescript would never tell me what is wrong with my theme because it infers the typings from whatever I write as a return value.

Preventing this type of errors is the whole point of using Typescript. If I simply need to be careful, then why am I using Typescript in the first place?

In `1.3.2`, the typings are too permissives for their own good. It will create problems for users.

Returned values should always be either `SystemStyleObject` or `SystemStyleObjectRecord`. Nothing else should be allowed.

⠀
⠀
⠀

## `#5` Consistency & Intent

This is a very minor point about how some typings are defined in `1.3.2`.

```typescript
//
// theme.types.d.ts
//
export interface SystemStyleObjectRecord {
   [key: string]: StyleObjectOrFn; // <--------
}
```

```typescript
//
// @chakra-ui/styled-system/dist/types/types.d.ts
//
export declare type CSSObject = RecursiveCSSObject<CSSWithMultiValues>;
export declare type SystemStyleObject = CSSObject; // <--------
export interface FunctionCSSInterpolation {
   (theme: Dict): CSSObject; // <--------
}
export declare type StyleObjectOrFn = CSSObject | FunctionCSSInterpolation; // <--------
```

I think it would be semantically better, and better readability, to have this instead:

```typescript
//
// theme.types.d.ts
//
export interface SystemStyleObjectRecord {
	// StyleObjectOrFn --> SystemStyleObjectOrFn
	//   because that's what is actually returned, either a function or SystemStyleObject
   [key: string]: SystemStyleObjectOrFn;
}
```

```typescript
//
// @chakra-ui/styled-system/dist/types/types.d.ts
//
export declare type CSSObject = RecursiveCSSObject<CSSWithMultiValues>;
export declare type SystemStyleObject = CSSObject;
export interface FunctionCSSInterpolation {
	// CSSObject --> SystemStyleObject
	//   while `CSSObject` and `SystemStyleObject` are the same type, `SystemStyleObject` is the type
	//   used elsewhere in the code to make the distinction very clear about what we're manipulating
   (theme: Dict): SystemStyleObject;
}

// StyleObjectOrFn --> SystemStyleObjectOrFn
//   because that's what is actually returned, either a function or SystemStyleObject
// CSSObject --> SystemStyleObject
//   better distinction about what we're manipulating
export declare type SystemStyleObjectOrFn = SystemStyleObject | FunctionCSSInterpolation;
```

There is no bug, it's really just a matter of consistency between intents.

Same principle could apply to type `SystemStyleObjectRecord`. By itself it doesn't really mean anything.

It could be easier to name it `SystemMultiStyleObject` to keep the same consistency established between `ComponentStyleConfig` and `ComponentMultiStyleConfig`.

⠀
⠀
⠀

## Conclusion

I've just completed the conversion of my typings to Chakra's `1.3.2`.

Putting aside the pain points, everything else works as intended.

It's possible there are other pain points I haven't stumbled upon, but for now that's all I found with the new typings from `1.3.2`.

**Comment by TimKolberger** [comment]
Thank you for your detailed feedback! 💖 

The typings for the theme were incrementally updated over the last releases to provide more typesafety while writing theme overrides.
Along the way we had some issues with the typings, because the theme object is not very simple 😅 
[Here is the changelog for furthere reference](https://github.com/chakra-ui/chakra-ui/blob/main/packages/theme/CHANGELOG.md).

I am feeling many of the mentioned pain points as well, because I am maintaining an extended theme as well. Because of that I am working on improving the typings.

Let me extract the actions points from your post:

##  1.-3.

* [ ] extract props type from `ThemingPropsThunk` into a standalone exported type
* [ ] add missing theme property to this new props type. 
   Inferring the current theme type would be awesome, but I think this is very hard to achieve, so I think this should be of type `ChakraTheme`.  Because the types are used to extend the theme with a random base theme, I think we cannot infer the types beforehand.
- [ ] The `framer-motion` warning in the CLI is a real bugger

PRs welcome 😇

> I tried to use the new Chakra CLI to see if I could generate the typings correctly, but documentation is a bit spotty. It doesn't say what am I supposed to import once the theme typings have been generated.

The Chakra CLI derives types from your custom theme to only provide autocomplete for the theme tokens on a component level.
So when your are typing  `<Input variant="` you get IntelliSense for all available variants for the Input component. It does not provide types to use them in your theme object. The snake would bite itself in the tail there 🐍 


## 4.

If this code snipped is a part of a `ComponentMultiStyleConfig` it should be valid afaik. I added some inline comments:

```js
  baseStyle: {
    // component part style object key of type `string`
    layParent: {},
    layChild: {
      // `brand1` could be a CSS selector like `.brand1::first-of-type` or smth simple as `span` so we allow all string values
      brand1: {
        color: 'red.100',
      },
      brand2: {
        // same goes for this object key: 
        // it could be a nested CSS selector 
        // e.g. this results in the following CSS: `brand2 subbrand1 { /* css properties */}` 
        // so this type needs to allow nested objects and a key of type `string`
        subbrand1: {
          color: 'blue.100',
        },
        subbrand2: {
          color: 'green.100',
        },
      },
    },
    preChild: {},
  },
```

---

> Returned values should always be either SystemStyleObject or SystemStyleObjectRecord. Nothing else should be allowed.

I think this is the case 🤔  Please let me know what I am missing there.

https://github.com/chakra-ui/chakra-ui/blob/main/packages/theme/src/theme.types.ts#L54-L71:

```ts
export interface SystemStyleObjectRecord {
  // part name
  [key: string]: StyleObjectOrFn // <---- Object
}

export interface ComponentSingleStyleConfig {
  baseStyle?: ThemingPropsThunk<SystemStyleObject> // <---- Object
  sizes?: Record<string, ThemingPropsThunk<SystemStyleObject>>
  variants?: Record<string, ThemingPropsThunk<SystemStyleObject>>
  defaultProps?: ComponentDefaultProps
}

export interface ComponentMultiStyleConfig {
  parts: string[]
  baseStyle?: ThemingPropsThunk<SystemStyleObjectRecord> // <---- Record for the parts
  sizes?: Record<string, ThemingPropsThunk<SystemStyleObjectRecord>>
  variants?: Record<string, ThemingPropsThunk<SystemStyleObjectRecord>>
  defaultProps?: ComponentDefaultProps
}
```

## 5.

The type names got me too a few times 😄  
I think we could rename those in the next major release _(in a few months)_

---

I hope I could answer a few questions of yours 🤗

**Reply by TheThirdRace** [reply]
Thank you for your very detailled answer.

Just to make sure, is your username `@hakunamatim` on Chakra's Discord channel? 

To answer your new questions...

> extract props type from ThemingPropsThunk into a standalone exported type

That seems about right

> add missing theme property to this new props type

Also right

> Inferring the current theme type would be awesome, but I think this is very hard to achieve, so I think this should be of type ChakraTheme. Because the types are used to extend the theme with a random base theme, I think we cannot infer the types beforehand.

Yep, that's what I was thinking. I'm far from being an expert in Typescript, but I couldn't find anything that would fit what I was looking for either. Too bad :(

> The framer-motion warning in the CLI is a real bugger

I know there's an issue opened for this in Chakra's github repository. I guess the problem will eventually be fixed.

> The Chakra CLI derives types from your custom theme to only provide autocomplete for the theme tokens on a component level.

So you're confirming this cannot be used for `theme only` styling right?

> If this code snipped is a part of a ComponentMultiStyleConfig it should be valid afaik. I added some inline comments:
> [...]
> >Returned values should always be either SystemStyleObject or SystemStyleObjectRecord. Nothing else should be allowed.
>
>I think this is the case 🤔 Please let me know what I am missing there.

No you're not missing anything. I never thought about how the custom CSS selectors were typed in this instance. 

You're completely right, even if I only saw custom CSS selectors, it could take the form of having properties with objects in it and have sub properties for a few levels deeper.

There's absolutely nothing we can do about this.

---

Again, thank your for your answers.

Hopefully, someone will eventually have time to make a PR for the first 2 points since they seem quickly fixable.

**Reply by TheThirdRace** [reply]
Actually, there's another situation with the types.

Sometimes, I need to execute another variant that is a function. But to do so, I need to type the variant as a function otherwise Typescript will complain that not all `SystemStyleObject` definitions are functions.

So no only `props` need to have its own type, but the whole function definition also need its own type.

If you look at my custom types in my first post, I was exporting a type for just the function part:

```typescript
export type StyleConfigFunction =
	| ((props: StyleConfigProps) => SystemStyleObject)
	| ((props: StyleConfigProps) => { [part: string]: SystemStyleObject })
```

**Reply by TimKolberger** [reply]
Yes, I am @hakunamatim on discord 👍 

I just opened a PR for the type export: https://github.com/chakra-ui/chakra-ui/pull/3622 :tada:

--- 

> Sometimes, I need to execute another variant that is a function. 

A thought on your circular dep issue: Maybe the [`import type` from TS3.8](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export) could help you with this. This should make `StyleConfigFunction` type obsolete 🤔 

---

> The framer-motion warning in the CLI is a real bugger

Kinda unreIated, but I opened a PR for the cli fix: https://github.com/chakra-ui/chakra-ui/pull/3621

> So you're confirming this cannot be used for theme only styling right?

Yes, the cli tokens command serves an other use case. It generates a type, which is only used internally to provide autocomplete in your IDE.

**Reply by TheThirdRace** [reply]
I will have a look at the `import type`, but it shouldn't have any impact on `StyleConfigFunction`. 

The circular dependency happens with the type of `theme` received in `ThemeComponentProps` vs the actual extended theme

**Reply by TimKolberger** [reply]
Yes, you are right - would have been great, but `import type` does not help.
